pkg/admission/kubequota/kubequota_admission.go

```diff
@@ -0,0 +1,359 @@
+				}
+			}
+
+			return NewKubeResourceQuota(configuration), nil
+		},
+	)
+}
+
+// NewKubeResourceQuota returns a new KubeResourceQuota admission plugin.
+func NewKubeResourceQuota(config *resourcequotaapi.Configuration) *KubeResourceQuota {
+	return &KubeResourceQuota{
+		Handler: admission.NewHandler(admission.Create, admission.Update),
+
+		userSuppliedConfiguration: config,
+
+		delegates: map[logicalcluster.Name]*stoppableQuotaAdmission{},
+	}
+}
+
+// KubeResourceQuota is an admission plugin that handles quota per logical cluster.
+type KubeResourceQuota struct {
+	*admission.Handler
+
+	// Injected/set via initializers
+	clusterWorkspaceInformer tenancyinformers.ClusterWorkspaceInformer
+	clusterWorkspaceLister   tenancylisters.ClusterWorkspaceLister
+	kubeClusterClient        kubernetes.ClusterInterface
+	kubeInformers            informers.SharedInformerFactory
+	quotaConfiguration       quota.Configuration
+	serverDone               <-chan struct{}
+
+	// Manually set
+	userSuppliedConfiguration *resourcequotaapi.Configuration
+
+	lock      sync.RWMutex
+	delegates map[logicalcluster.Name]*stoppableQuotaAdmission
+
+	once sync.Once
+}
+
+// ValidateInitialization validates all the expected fields are set.
+func (k *KubeResourceQuota) ValidateInitialization() error {
+	if k.clusterWorkspaceLister == nil {
+		return fmt.Errorf("missing clusterWorkspaceLister")
+	}
+	if k.kubeClusterClient == nil {
+		return fmt.Errorf("missing kubeClusterClient")
+	}
+	if k.kubeInformers == nil {
+		return fmt.Errorf("missing kubeInformers")
+	}
+	if k.quotaConfiguration == nil {
+		return fmt.Errorf("missing quotaConfiguration")
+	}
+	if k.serverDone == nil {
+		return fmt.Errorf("missing serverDone")
+	}
+
+	return nil
+}
+
+var _ admission.ValidationInterface = &KubeResourceQuota{}
+var _ = initializers.WantsKcpInformers(&KubeResourceQuota{})
+var _ = initializer.WantsExternalKubeInformerFactory(&KubeResourceQuota{})
+var _ = initializers.WantsKubeClusterClient(&KubeResourceQuota{})
+var _ = initializer.WantsQuotaConfiguration(&KubeResourceQuota{})
+var _ = initializers.WantsServerShutdownChannel(&KubeResourceQuota{})
+
+// Validate gets or creates a resourcequota.QuotaAdmission plugin for the logical cluster in the request and then
+// delegates validation to it.
+func (k *KubeResourceQuota) Validate(ctx context.Context, a admission.Attributes, o admission.ObjectInterfaces) error {
+	k.once.Do(func() {
+		m := newClusterWorkspaceDeletionMonitor(k.clusterWorkspaceInformer, k.stopQuotaAdmissionForCluster)
+		go m.Start(k.serverDone)
+	})
+
+	cluster, err := genericapirequest.ValidClusterFrom(ctx)
+	if err != nil {
+		return err
+	}
+
+	delegate, err := k.getOrCreateDelegate(cluster.Name)
+	if err != nil {
+		return err
+	}
+
+	// hackyA := &hackyAdmissionAttributesForBuiltInTypes{
+	// 	Attributes: a,
+	// }
+
+	return delegate.Validate(ctx, a, o)
+}
+
+// TODO(ncdc): find a way to support the special evaluators for pods/pvcs/services.
```

Comment:
Important? I know you said you'd changed the approach here ...

Reply:
Comment removed in next push